### DISCUSSION TITLE
How to properly use BytesMut with UnixDatagram recv/recv_from?

### DISCUSSION THREAD

**Discussion by xonatius** [tokio-rs/tokio#2741] [General] [answered]
I was wondering what is the correct way of `recv`-ing data into [BytesMut](https://docs.rs/bytes/0.5.6/bytes/struct.BytesMut.html) from [UnixDatagram](https://docs.rs/tokio/0.2.22/tokio/net/struct.UnixDatagram.html). The only safe way to do it seems to be `resize`-ing the buffer to its capacity and then `truncate`-ing till the received length, but that will add a cost of initializing data which is not going to be read anyway. The other approach seems to be to use `unsafe` block and use `set_len` to buf capacity and then `set_len` to the actual amount of bytes read. The later approach avoids unnecessary initialization, but is it safe to do?

**Comment by Darksonn** [comment, marked as answer]
This is generally a problem with all IO sources. If you don't want to zero the memory first, the correct way to do it is to call [`BufMut::bytes_mut`][1] to obtain a reference into the extra capacity, then unsafely cast it to an `&mut [u8]`, which you can then give to the datagram. You can then call [`BufMut::advance_mut`][2] after you know how much data was written into the slice.

[1]: https://docs.rs/bytes/0.5/bytes/buf/trait.BufMut.html#tymethod.bytes_mut
[2]: https://docs.rs/bytes/0.5/bytes/buf/trait.BufMut.html#tymethod.advance_mut

**Reply by xonatius** [reply]
Thanks! As for `set_len` - is there any issue with using that approach?

**Reply by Darksonn** [reply]
The `advance_mut` method just calls `set_len` internally.

**Reply by xonatius** [reply]
> the correct way to do it is to call [`BufMut::bytes_mut`](https://docs.rs/bytes/0.5/bytes/buf/trait.BufMut.html#tymethod.bytes_mut) to obtain a reference into the extra capacity, then unsafely cast it to an `&mut [u8]`,

Hm, that did not work for me, as I can't spawn a task with `recv` then as `*mut [u8]` is not send-able:

``` error[E0277]: `*mut [u8]` cannot be sent between threads safely ```

**Reply by Darksonn** [reply]
That error message does not seem related to `BufMut::bytes_mut`.
